web/modules/custom/weather_block/src/Services/GetSetCityUser.php

```diff
@@ -0,0 +1,128 @@
+<?php
+
+namespace Drupal\weather_block\Services;
+
+use Drupal\Core\Database\Connection;
+use Drupal\Core\Http\ClientFactory;
+use Drupal\Core\Logger\LoggerChannelFactoryInterface;
+use GuzzleHttp\Exception\ClientException;
+
+/**
+ * Save cities for users.
+ */
+class GetSetCityUser {
+
+  /**
+   * Initialize a database property.
+   */
+  public function __construct(protected Connection $database, protected LoggerChannelFactoryInterface $logger, protected ClientFactory $httpClient) {
+  }
+
+  /**
+   * Saves selected cities for a user in the database.
+   *
+   * @throws \Exception
+   */
+  public function saveCitiesForUser($userId, $cities): void {
+    $this->database->delete('weather_block_user_cities')
+      ->condition('user_id', $userId)
+      ->execute();
+
+    foreach ($cities as $city) {
+      $this->database->insert('weather_block_user_cities')
+        ->fields([
+          'user_id' => $userId,
+          'city' => $city,
+        ])
+        ->execute();
+    }
+  }
+
+  /**
+   * Retrieves saved cities for a user from the database.
+   */
+  public function getSavedCitiesForUser($userId): array {
+    $query = $this->database->select('weather_block_user_cities', 'w');
+    $query->fields('w', ['city']);
+    $query->condition('w.user_id', $userId);
+    $result = $query->execute()->fetchAll();
+    $cities = [];
+    foreach ($result as $row) {
+      $cities[] = $row->city;
+    }
+    return $cities;
+  }
+
+  /**
+   * Saves weather data for a city in the database.
+   */
+  public function saveWeatherDataForCity($city, array $weatherData): void {
+    $existingRecord = $this->database->select('weather_block_user_cities', 'w')
+      ->fields('w')
+      ->condition('city', $city)
+      ->execute()
+      ->fetchAssoc();
+
+    if ($existingRecord) {
+      $this->database->update('weather_block_user_cities')
+        ->fields([
+          'weather_data' => serialize($weatherData),
+          'cache_expire' => time() + 60 * 60,
+        ])
+        ->condition('city', $city)
+        ->execute();
+    }
+    else {
+      $this->database->insert('weather_block_user_cities')
+        ->fields([
+          'user_id' => \Drupal::currentUser()->id(),
+          'city' => $city,
+          'weather_data' => serialize($weatherData),
+          'cache_expire' => time() + 60 * 60,
+        ])
+        ->execute();
+    }
+  }
+
+  /**
+   * Private function for getting array with temperature.
+   */
+  public function getDataFromApi(string $cities, string $api_key): array {
```

Comment:
If the service is responsible for fetching data, then it should handle $api_key by itself.
It will be just annoying to do the same thing for getting $api_key in every place you want to use that method.